### PR TITLE
Creates the SQL database connection and all the required tables.

### PR DESCRIPTION
**Pull request by Cayle**
This file sets up the connection to the sql database and also creates all tables required for the project.

### CODE REVIEW

File: capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+import javax.servlet.ServletContextEvent;
+import javax.servlet.ServletContextListener;
+import javax.servlet.annotation.WebListener;
+import javax.sql.DataSource;
+
+@WebListener("Creates a connection pool that is stored in the Servlet's context for later use.")
+public class ConnectionPoolContextListener implements ServletContextListener {
+
+    // Saving credentials in environment variables is convenient, but not secure - consider a more
+    // secure solution such as https://cloud.google.com/kms/ to help keep secrets safe.
```
Comment (ljfranklin) [inline capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java:34]: 👍 this is a nice stretch goal

File: capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    private static final String CLOUD_SQL_CONNECTION_NAME = System.getenv("CLOUD_SQL_CONNECTION_NAME");
+    private static final String DB_USER = System.getenv("DB_USER");
+    private static final String DB_PASS = System.getenv("DB_PASS");
+    private static final String DB_NAME = System.getenv("DB_NAME");
+
+    private DataSource createConnectionPool() {
+        // [START cloud_sql_postgres_servlet_create]
+        // The configuration object specifies behaviors for the connection pool.
```
Comment (ljfranklin) [inline capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java:42]: Let's remove all these comments, the code is easily readable without them. If you feel like you still want additional explanation consider better variable names instead, e.g. `config` -> `dbConfig`.

File: capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        return pool;
+    }
+
+    private void createUniversityTable(DataSource pool) throws SQLException {
+        // Safely attempts to create the university table schema.
+        try (Connection conn = pool.getConnection()) {
+            String stmt =
+                "CREATE TABLE IF NOT EXISTS university ( "
+                    + "id INT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY, name TEXT, state TEXT );";
+            try (PreparedStatement createTableStatement = conn.prepareStatement(stmt); ) {
```
Comment (ljfranklin) [inline capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java:69]: Remove the extra `; `

File: capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        DataSource pool = new HikariDataSource(config);
+        // [END cloud_sql_postgres_servlet_create]
+        return pool;
+    }
+
+    private void createUniversityTable(DataSource pool) throws SQLException {
+        // Safely attempts to create the university table schema.
+        try (Connection conn = pool.getConnection()) {
+            String stmt =
+                "CREATE TABLE IF NOT EXISTS university ( "
```
Comment (ljfranklin) [inline capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java:67]: Nit: move all the SQL statements into constants towards the top of the file. The table defs are the key piece of info in this file so optimize for a reader seeing those lines as soon as they open the file.

**Comment by Cayle** [comment]
Made all the requested changes.

File: capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+  public void contextInitialized(ServletContextEvent event) {
+    // This function is called when the application starts and will safely create a connection pool
+    // that can be used to connect to.
+    ServletContext servletContext = event.getServletContext();
+    DataSource pool = (DataSource) servletContext.getAttribute("my-pool");
+    if (pool == null) {
+      pool = createConnectionPool();
+      servletContext.setAttribute("my-pool", pool);
+    }
+    //After the connection pool has been set up, we try to create our tables in the database.
```
Comment (ljfranklin) [inline capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java:157]: Nit: remove these comments too

File: capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+  // Saving credentials in environment variables is convenient, but not secure - consider a more
+  // secure solution such as https://cloud.google.com/kms/ to help keep secrets safe.
+  private static final String CLOUD_SQL_CONNECTION_NAME = System.getenv(
+    "CLOUD_SQL_CONNECTION_NAME"
+  );
+  private static final String DB_USER = System.getenv("DB_USER");
+  private static final String DB_PASS = System.getenv("DB_PASS");
+  private static final String DB_NAME = System.getenv("DB_NAME");
+
+  private static final String UNIVERSITY_TABLE_SQL_STATEMENT =
```
Comment (ljfranklin) [inline capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java:43]: Fine to submit this as is, but consider a follow-up CL to move all these statements into a `.sql` file which the Java file loads on startup. Same functionality but it'll be a lot easier to edit compared to editing string literals in Java.

File: capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java (hunk 5 above)
Comment (Cayle) [reply capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java]: Okay, removed it.

File: capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java (hunk 6 above)
Comment (Cayle) [reply capstone/src/main/java/com/google/sps/servlets/ConnectionPoolContextListener.java:43]: Okay, I'll look into how to do that.

**Pull request merged**